packages/react-sdk/src/withOptimizely.tsx

```diff
@@ -17,36 +17,44 @@ import * as React from 'react'
 import { Subtract } from 'utility-types'
 
 import { OptimizelyContextConsumer } from './Context'
-import { ReactSDKClient } from './client';
+import { ReactSDKClient } from './client'
 
 export interface WithOptimizelyProps {
   optimizely: ReactSDKClient | null
   optimizelyReadyTimeout: number | undefined
   isServerSide: boolean
 }
 
-export function withOptimizely<P extends WithOptimizelyProps>(
+export function withOptimizely<P extends WithOptimizelyProps, R>(
   Component: React.ComponentType<P>,
-): React.ComponentType<Subtract<P, WithOptimizelyProps>> {
-  return class WithOptimizely extends React.Component<Subtract<P, WithOptimizelyProps>> {
-    render() {
-      return (
-        <OptimizelyContextConsumer>
-          {(value: {
-            optimizely: ReactSDKClient
-            isServerSide: boolean
-            timeout: number | undefined
-          }) => (
-            // @ts-ignore
-            <Component
-              {...this.props}
-              optimizelyReadyTimeout={value.timeout}
-              optimizely={value.optimizely}
-              isServerSide={value.isServerSide}
-            />
-          )}
-        </OptimizelyContextConsumer>
-      )
-    }
+): React.ForwardRefExoticComponent<
+  React.PropsWithoutRef<Subtract<P, WithOptimizelyProps>> & React.RefAttributes<R>
+> {
+  const withOptimizely: React.RefForwardingComponent<
+    R,
+    Subtract<P, WithOptimizelyProps>
+  > = function withOptimizely(props, ref) {
+    return (
+      // Note: Casting props to P is necessary because of this TypeScript issue:
+      // https://github.com/microsoft/TypeScript/issues/28884
+      <OptimizelyContextConsumer>
+        {(value: {
+          optimizely: ReactSDKClient
+          isServerSide: boolean
+          timeout: number | undefined
+        }) => (
+          <Component
+            {...props as P}
+            optimizely={value.optimizely}
+            optimizelyReadyTimeout={value.timeout}
+            isServerSide={value.isServerSide}
+            ref={ref}
+          />
+        )}
+      </OptimizelyContextConsumer>
+    )
   }
+  withOptimizely.displayName = `withOptimizely(${Component.displayName ||
```

Comment:
Checkout the helper I wrote in the segment-js repo to forward refs and hoist non-react statics.

I think copy/pasting that here would be useful.

https://github.com/optimizely/segment-js/blob/master/src/decorators.js#L14-L29

Reply:
@jamesopti I tried using this helper function, but I'm finding this to make the types a lot more challenging. Would you be OK if we defer this until later? FWIW, this `withOptimizely` is the only HoC that would be using this helper function at the moment.